build_tools/scripts/git/update_runtime_submodules.sh

```diff
@@ -14,4 +14,9 @@ SCRIPT_DIR="$(dirname -- "$( readlink -f -- "$0"; )")"
 
 readarray -t RUNTIME_SUBMODULES < "${SCRIPT_DIR}/runtime_submodules.txt"
 
-git submodule sync && git submodule update --init ${RUNTIME_SUBMODULES[@]}
+# Trim \r characters on Windows (-t above only removes \n).
+for i in "${!RUNTIME_SUBMODULES[@]}"; do
+  RUNTIME_SUBMODULES[$i]=$(echo ${RUNTIME_SUBMODULES[$i]} | tr -d '\r')
+done
```

Comment:
This item by item processing is noisy: 
![image](https://user-images.githubusercontent.com/4010439/199596330-8a93bf9e-ebf1-48ca-bf78-b03743a5cbd0.png)

Maybe someone can suggest a one-liner that has the same function?

Reply:
`-d delim  Use DELIM to terminate lines, instead of newline`? But that means we'd have to switch on whether it's windows.

Should we be getting windows newlines here at all? I can't remember how the git newlines magic works. Is it that it stores as unix and then when checked out on windows automagically makes them have carriage returns?

For a shorter processing, you could do it on the original file instead of after parsing the array

```shell
"${SCRIPT_DIR}/runtime_submodules.txt" | tr -d '\r' | readarray -t RUNTIME_SUBMODULES
```

Reply:
> `-d delim Use DELIM to terminate lines, instead of newline`? But that means we'd have to switch on whether it's windows.

`\r\n` is two characters, while `readarray` says `-d     The first character of delim is used to terminate each input line, rather than newline.` :/

> Should we be getting windows newlines here at all? I can't remember how the git newlines magic works. Is it that it stores as unix and then when checked out on windows automagically makes them have carriage returns?

See the section on `core.autocrlf` in https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration. I looked for a way to set line endings within GitHub Actions and just found https://github.com/actions/checkout/issues/135 

There's a workaround there, but it's pretty verbose:
```yaml
    steps:
      - name: Set git to use LF
        run: |
          git config --global core.autocrlf false
          git config --global core.eol lf

      - uses: actions/checkout@v2
```

> For a shorter processing, you could do it on the original file instead of after parsing the array
> 
> ```shell
> "${SCRIPT_DIR}/runtime_submodules.txt" | tr -d '\r' | readarray -t RUNTIME_SUBMODULES
> ```

I tried this style of piping, but `readarray` wants stdin to provide a file name, not the contents of a file :/

Reply:
That doesn't seem right. Here we are explicitly passing the file via stdin also (that's what the `<` does). Also contradicts the docs: " Read lines from the standard input into an indexed array variable."

Reply:
¯\\\_(ツ)\_/¯
```
echo "${SCRIPT_DIR}/runtime_submodules.txt" | tr -d '\r' | readarray -t RUNTIME_SUBMODULES
echo ${RUNTIME_SUBMODULES}

.\build_tools\scripts\git\update_runtime_submodules.sh: line 23: RUNTIME_SUBMODULES: unbound variable
```

Reply:
You want `cat` not echo

Reply:
Ugh but even with that you're right. Apparently you need process substitution. Thanks bash: https://unix.stackexchange.com/questions/136206/readarray-or-pipe-issue

Reply:
`readarray -t RUNTIME_SUBMODULES < <(cat build_tools/scripts/git/runtime_submodules.txt  | tr -d '\r')`

What a stupid programming language

Reply:
That works! Thanks much :)